embassy-stm32/src/spi/mod.rs

```diff
@@ -574,36 +574,90 @@ impl<'d, T: Instance, Tx, Rx> Spi<'d, T, Tx, Rx> {
         self.transfer_inner(data, data).await
     }
 
-    pub fn blocking_write<W: Word>(&mut self, words: &[W]) -> Result<(), Error> {
+    fn configure_blocking<W: Word>(&mut self) -> ConfiguredBlocking<'_, W> {
+        flush_rx_fifo(T::REGS);
         self.set_word_size(W::WORDSIZE);
+        ConfiguredBlocking {
+            regs: T::REGS,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn blocking_write<W: Word>(&mut self, words: &[W]) -> Result<(), Error> {
+        self.configure_blocking().blocking_write(words)
+    }
+
+    pub fn blocking_read<W: Word>(&mut self, words: &mut [W]) -> Result<(), Error> {
+        self.configure_blocking().blocking_read(words)
+    }
+
+    pub fn blocking_transfer_in_place<W: Word>(&mut self, words: &mut [W]) -> Result<(), Error> {
+        self.configure_blocking().blocking_transfer_in_place(words)
+    }
+
+    pub fn blocking_transfer<W: Word>(&mut self, read: &mut [W], write: &[W]) -> Result<(), Error> {
+        self.configure_blocking().blocking_transfer(read, write)
+    }
+}
+
+impl<'d, T: Instance, Tx, Rx> Drop for Spi<'d, T, Tx, Rx> {
+    fn drop(&mut self) {
+        unsafe {
+            self.sck.as_ref().map(|x| x.set_as_disconnected());
+            self.mosi.as_ref().map(|x| x.set_as_disconnected());
+            self.miso.as_ref().map(|x| x.set_as_disconnected());
+        }
+    }
+}
+
+struct ConfiguredBlocking<'a, W: Word> {
```

Comment:
Could you do this without a separate struct? It adds a bit of boilerplate and indirection, and since only a private API, the added type safety is not that worth it IMO.